p2p/transport/webrtc/udp_mux.go

```diff
@@ -0,0 +1,341 @@
+package libp2pwebrtc
+
+import (
+	"errors"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pion/ice/v2"
+	"github.com/pion/logging"
+	"github.com/pion/stun"
+)
+
+type candidateAddr struct {
+	raddr *net.UDPAddr
+	ufrag string
+}
+
+// UDPMuxNewAddr is an implementation of the interface
+type UDPMuxNewAddr struct {
+	params ice.UDPMuxParams
+
+	closedChan chan struct{}
+	closeOnce  sync.Once
+
+	// connsIPv4 and connsIPv6 are maps of all udpMuxedConn indexed by ufrag|network|candidateType
+	connsIPv4, connsIPv6 map[string]*udpMuxedConn
+
+	addressMapMu sync.RWMutex
+	addressMap   map[string]*udpMuxedConn
+
+	// buffer pool to recycle buffers for net.UDPAddr encodes/decodes
+	pool *sync.Pool
+
+	mu sync.Mutex
+
+	newAddrChan chan candidateAddr
+	newAddrs    map[*net.UDPAddr]struct{}
+}
+
+const maxAddrSize = 512
+const receiveMTU = 8192
+
+// NewUDPMuxNewAddr creates an implementation of UDPMux
+func NewUDPMuxNewAddr(params ice.UDPMuxParams, newAddrChan chan candidateAddr) *UDPMuxNewAddr {
+	if params.Logger == nil {
+		params.Logger = logging.NewDefaultLoggerFactory().NewLogger("ice")
+	}
+
+	m := &UDPMuxNewAddr{
+		addressMap: map[string]*udpMuxedConn{},
+		params:     params,
+		connsIPv4:  make(map[string]*udpMuxedConn),
+		connsIPv6:  make(map[string]*udpMuxedConn),
+		closedChan: make(chan struct{}, 1),
+		pool: &sync.Pool{
+			New: func() interface{} {
+				// big enough buffer to fit both packet and address
+				return newBufferHolder(receiveMTU + maxAddrSize)
+			},
+		},
+		newAddrChan: newAddrChan,
+		newAddrs:    make(map[*net.UDPAddr]struct{}),
+	}
+
+	go m.connWorker()
+
+	return m
+}
+
+// LocalAddr returns the listening address of this UDPMuxNewAddr
+func (m *UDPMuxNewAddr) LocalAddr() net.Addr {
+	return m.params.UDPConn.LocalAddr()
+}
+
+// GetConn returns a PacketConn given the connection's ufrag and network
+// creates the connection if an existing one can't be found
+func (m *UDPMuxNewAddr) GetConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.IsClosed() {
+		return nil, io.ErrClosedPipe
+	}
+
+	if conn, ok := m.getConn(ufrag, isIPv6); ok {
+		return conn, nil
+	}
+
+	c := m.createMuxedConn(ufrag)
+	go func() {
```

Comment:
Any way to avoid spawning a new Go routine for every connection?
If not, this will need careful consideration during shutdown to make sure that _all_ Go routines have returned before we return from `Close`.